url3f8d6440dd_huiwq1990_CognativeModel_tgz-pJ8-clarion_samples_HelloWorldJ8/src/clarion/system/RuleCollection.java

```diff
@@ -0,0 +1,862 @@
+	{
+		HashMap<Object, HashMap<Object,LinkedList <AbstractRule>>> rc = get(R.getAction().getID());
+		if (rc == null || R.getCondition() == null)
+			return null;
+		HashSet <AbstractRule> candidates = new HashSet<AbstractRule> ();
+		boolean flag = false;
+		for(Dimension d : R.getCondition().values())
+		{
+			HashMap<Object,LinkedList <AbstractRule>> rrc = rc.get(d.getID());
+			if(rrc == null)
+				return null;
+			for(Value v : d.values())
+			{
+				if(v.isFullyActivated())
+				{
+					if(!flag)
+					{
+						candidates.addAll(rrc.get(v.getID()));
+						flag = true;
+					}
+					else
+					{
+						LinkedList <AbstractRule> rem = rrc.get(v.getID());
+						HashSet <AbstractRule> cut = new HashSet <AbstractRule> ();
+						cut.addAll(candidates);
+						for(AbstractRule a : rem)
+						{
+							if(candidates.contains(a))
+							{
+								cut.remove(a);
+							}
+						}
+						
+						candidates.removeAll(cut);
+					}
+				}
+			}
+		}
+		for(Iterator <AbstractRule> i = candidates.iterator(); i.hasNext();)
+		{
+			AbstractRule a = i.next();
+			if(a.getCondition().equals(R.getCondition()) || !a.getCondition().equalsKeys(R.getCondition()))
+				i.remove();
+		}
+		if(candidates.size() == 0)
+			return null;
+		else
+			return candidates;
+	}
+	
+	/**
+	 * Checks to see if the collection contains the specified rule.
+	 * @param rule The rule object you wish to check for within this collection.
+	 * @return True if the collection contains the rule specified, otherwise false.
+	 */
+	public boolean contains (AbstractRule rule)
+	{
+		if(get(rule.rCondition, rule.rAction) != null)
+			return true;
+		return false;
+	}
+	
+	/**
+	 * Checks to see if the collection contains a rule that covers the specified rule.
+	 * @param rule The rule object you wish to check within this collection.
+	 * @return True if the collection contains a rule that covers the rule specified, 
+	 * otherwise false.
+	 */
+	public boolean containsCover (AbstractRule rule)
+	{
+		if(getCover(rule) != null)
+			return true;
+		return false;
+	}
+
+	/**
+	 * Adds a rule into the rule collection. If the specified rule is covered by any rules 
+	 * currently in this collection, then the rule will be added to the collection of child 
+	 * rules for that rule. If the specified rule covers any rules in this collection, those rules
+	 * will be added to the child rules of the specified rule and the specified rule will 
+	 * replace those rules in this collection. If the specified rule is equal to any rules 
+	 * currently in this collection the rule will NOT be added. If the condition of the specified
+	 * rule is null, it will be added to the "null" slot of the rule collection. If there is already
+	 * a condition-less rule in the "null" slot, it will be REPLACED by the specified rule. 
+	 * @param R The rule to add.
+	 * @return True if the rule was successfully added, otherwise false.
+	 */
+	public boolean add (AbstractRule R)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 39. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.